### ISSUE TITLE
Nano S not recognized

### ISSUE THREAD

**Issue by JPUnix** [fyookball/electrum#149]
Trying to install Electron Cash 2.9.3 under Windows 7 using a Ledger Nano . I select "Standard wallet" -> "Use a hardware device". Nano S is plugged in and the BCC app is open. I get the following error message:

No hardware wallet support found on your system.
Please install the relevant libraries (eg python-trezor for Trezor)

**Comment by molecular** [comment]
I found my hw.1 device, but can't get it to work ("no hardware wallet support found", neither in electrum nor in electron-cash. Installed packages "btchip-udev" and "python2-btchip" on my system to no avail.

Does your Nano S work with electrum at all?

**Comment by molecular** [comment]
I dug in and found my hw.1 is likely not supported by python2-btchip any more ;-(

I say this because in the code starting here: https://github.com/LedgerHQ/btchip-python/blob/5fdf6fae1b33be2d9248af8ee0eda14da0a1a0c3/btchip/btchipComm.py#L173 the usb id my hw.1 device has (2581:1b7c) isn't listed.

I'm afraid I can't help without a device.

**Comment by molecular** [comment]
Trying to get the firmware of my hw.1 updated since it's not supported any more by python-btchip

**Comment by molecular** [comment]
Ok, I have a working btchip device (thanks to ledger for giving me some)!

I setup a wallet using their web-based wallet and was able to create the same one using electron-cash.

@JPUnix do you have btchip-python installed and if so, what version?

to see if btchip is installed in your python system, you can put the folling into a file test_btchip.py

```
from btchip.btchipComm import HIDDongleHIDAPI, DongleWait
from btchip.btchip import btchip
from btchip.btchipUtils import compress_public_key,format_transaction, get_regular_input_script, get_p2sh_input_script
from btchip.bitcoinTransaction import bitcoinTransaction
from btchip.btchipFirmwareWizard import checkFirmware, updateFirmware
from btchip.btchipException import BTChipException
print "btchip seems to be installed"
```

and then run it:

`#> python2.7 test_btchip.py`

**Comment by molecular** [comment]
@JPUnix on second thought: maybe you can just run electron cash from the terminal and look at the output? If btchip is not installed, it should look like:

```
  ~/bitcoin/Electron Cash-2.9.3\> ./electron-cash
No systemtrayicon available
Traceback (most recent call last):
  File "./electron-cash", line 377, in <module>
    d.init_gui(config, plugins)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/daemon.py", line 283, in init_gui
    self.gui.main()
  File "/home/nick/bitcoin/Electron Cash-2.9.3/gui/qt/__init__.py", line 215, in main
    if not self.start_new_window(path, self.config.get('url')):
  File "/home/nick/bitcoin/Electron Cash-2.9.3/gui/qt/__init__.py", line 171, in start_new_window
    wallet = self.daemon.load_wallet(path, None)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/daemon.py", line 220, in load_wallet
    wallet = Wallet(storage)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/wallet.py", line 1747, in __new__
    wallet = WalletClass(storage)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/wallet.py", line 1551, in __init__
    Deterministic_Wallet.__init__(self, storage)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/wallet.py", line 1395, in __init__
    Abstract_Wallet.__init__(self, storage)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/wallet.py", line 105, in __init__
    self.load_keystore()
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/wallet.py", line 1518, in load_keystore
    self.keystore = load_keystore(self.storage, 'keystore')
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/keystore.py", line 646, in load_keystore
    k = hardware_keystore(d)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/keystore.py", line 630, in hardware_keystore
    return constructor(d)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/plugins.py", line 170, in dynamic_constructor
    return self.get_plugin(name).keystore_class(d)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/plugins.py", line 178, in get_plugin
    self.load_plugin(name)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/lib/plugins.py", line 102, in load_plugin
    p = loader.load_module(full_name)
  File "/usr/lib/python2.7/pkgutil.py", line 246, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/home/nick/bitcoin/Electron Cash-2.9.3/plugins/ledger/qt.py", line 12, in <module>
    from btchip.btchipPersoWizard import StartBTChipPersoDialog
ImportError: No module named btchip.btchipPersoWizard
```

**Comment by molecular** [comment]
I can reproduce this on a windows machine at work.

I managed to capture the following output by running electron-cash-2.9.3.exe from a terminal:

```
$ /cygdrive/c/Users/slf/Downloads/electron-cash-2.9.3.exe
Traceback (most recent call last):
  File "C:\cygwin64\tmp\_MEI62122\lib\plugins.py", line 150, in get_hardware_support
    p = self.get_plugin(name)
  File "C:\cygwin64\tmp\_MEI62122\lib\plugins.py", line 178, in get_plugin
    self.load_plugin(name)
  File "C:\cygwin64\tmp\_MEI62122\lib\plugins.py", line 102, in load_plugin
    p = loader.load_module(full_name)
  File "Z:\home\ubuntu\gitwork\electrum\contrib\build-wine\build\deterministic\out00-PYZ.pyz\pkgutil", line 246, in load_module
  File "C:\cygwin64\tmp\_MEI62122\plugins\ledger\qt.py", line 12, in <module>
    from btchip.btchipPersoWizard import StartBTChipPersoDialog
ImportError: No module named btchip.btchipPersoWizard
```

I compared to electrum and the btchip device works there.

Seems we're missing btchip python dependency.

When running the .exe files, temporary folders are created and the python program is run from there (I assume). I checked those folders, neither in the electron-cash nor in the electrum case do they contain any files with "btchip" in their name. I'm confused.

**Comment by fyookball** [comment]
you need to include the packages when you build the exe...before you run it.

**Comment by molecular** [comment]
I didn't build the exe myself. Used the one downloaded from your page.

**Comment by fyookball** [comment]
so we need to include this btchip python library in packages in the next build we release.  Will do.

**Comment by willyloman** [comment]
@fyookball @molecular Thanks for looking into this issue. Looks like we've pin-pointed the cause of the issue to a missing btchip .exe? Would it be possible to cut a new release soon with the btchip lib included? Looks like there are 2 other outstanding issues #153, #126 which might be patched as well.

**Comment by bderose** [comment]
@fyookball Thanks for the 2.9.4 release, looks like we're making progress on this issue. However, on macOS 10.13.1 I am unable to restore a multi-sig wallet from my hardware devices. When I go `File -> New/Restore -> Multi-Signature -> Hardware Wallet` Electron Cash seems to hang on this screen:
![screen shot 2017-11-14 at 11 28 40 am](https://user-images.githubusercontent.com/31048768/32800560-852ae704-c92f-11e7-9aaa-d45ea7f671de.png)

Is there a log file I can copy/paste output from to help diagnose this?

**Comment by willyloman** [comment]
@fyookball Thanks for the 2.9.4 release, looks like we're making progress on this issue. However, on macOS 10.13.1 I am now unable to restore a multisig wallet from a Ledger S Nano hardware device in `Bitcoin Cash` mode.

Repro steps: `File -> New/Restore -> Next -> Multi-signature wallet -> 2of3 -> Hardware device` hangs the program at this screen:
![screen shot 2017-11-14 at 11 28 40 am](https://user-images.githubusercontent.com/30201235/32801081-da790ec4-c930-11e7-9ac7-1cf6e3a4b0cd.png)

Is there a log file I can copy/paste output from to help diagnose the issue?

**Comment by Danconnolly** [comment]
```
Traceback (most recent call last):
  File "/home/user/Projects/electroncash/lib/plugins.py", line 150, in get_hardware_support
    p = self.get_plugin(name)
  File "/home/user/Projects/electroncash/lib/plugins.py", line 178, in get_plugin
    self.load_plugin(name)
  File "/home/user/Projects/electroncash/lib/plugins.py", line 102, in load_plugin
    p = loader.load_module(full_name)
  File "/usr/lib64/python2.7/pkgutil.py", line 246, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/home/user/Projects/electroncash/plugins/ledger/qt.py", line 12, in <module>
    from btchip.btchipPersoWizard import StartBTChipPersoDialog
  File "/usr/lib/python2.7/site-packages/btchip/btchipPersoWizard.py", line 23, in <module>
    from PyQt4.QtGui import QDialog, QMessageBox
ImportError: cannot import name QDialog
```
Looks like the python2 btchip depends on QT4, we've moved to QT5.

**Comment by willyloman** [comment]
Thanks for the stacktrace @Danconnolly. What command are you issuing to give that error?

What’s the best path forward to fix this? We could fork btchip to use QT5, but I’m hesitant to lock us into a fork of brchip that might miss future updates. Seems a little obnoxious to make our dependencies update in response to a regression of our own. 

What was the rational for updating to QT5 in this repo? Would it be possible or easy to rollback to QT4?

Seems like this feature might warrant an integration test given the frequency we’ve been bitten by it.

**Comment by Danconnolly** [comment]
I was creating a new wallet, as you were. Unfortunately, I dont have an existing wallet linked to my nano, existing wallets may still work ok.

I was looking into the python3 version of btchip, hoping for Qt5, but it also seems to be dependent on QT4 (there's a plan for Electron Cash to move to python3 eventually).

I dont think its a good idea to move back to QT4. 

At the moment, it may be the initialization dialog code from btchip that use Qt, I'm thinking of just importing that code from btchip and updating to QT5. That way, we can still use the rest of the btchip library and remove our update when they update.

**Comment by willyloman** [comment]
Is there a vulnerability or deficiency in QT4 that’s preventing us from reverting those changes? Or are there features in QT5 that we really want/need? Given that this is a breaking change, it seems reasonable to discuss a downgrade to QT4 so long as it’s safe. 
 
Otherwise we could make the change to btchip, but there’s no guarantees if/when those changes get accepted. 

I’m no python expert, but would it also be possible to include QT4 in the electron cash build? Seems hacky and might get us into more trouble down the line, but just throwing that out as an option.

**Comment by Danconnolly** [comment]
Created issue #269 for the QT4 vs QT5 problem because this has a different root cause from the origin of this issue.

**Comment by roybadami** [comment]
I still see a missing module in 2.9.4 on OS X

`ImportError: No module named btchip.btchipPersoWizard`

**Comment by roybadami** [comment]
Also, looks like trezorlib.* (aka python-trezor) is missing from the Mac build, too.  (You don't see the import error unless you edit trezor.py to print it, since trezor.py silently catches the error by default.)

**Comment by willyloman** [comment]
Has anyone been able to verify this is fixed on v3.0? I'm using the Linux build of 3.0 and it doesn't seem to be picking up my Ledger S Nano over USB:

<img width="605" alt="screen shot 2017-12-12 at 2 28 24 pm" src="https://user-images.githubusercontent.com/30201235/33912182-b6b433c2-df49-11e7-824e-686106e8f432.png">

Props for the additional copy on the dialog though, will hopefully save folks some time.

**Comment by marceloneil** [comment]
@willyloman can you try this on 3.1.1? I've tested the ledger on all distributions and it should work.

**Comment by marceloneil** [comment]
@fyookball this should work (although I'm a bit confused as to why @willyloman didn't detect his nano s, since that should have worked in 3.0)

**Comment by willyloman** [comment]
fixed on 3.1.1, thanks @marceloneil

**Comment by ekseesse** [comment]
Hi, I have the same problem as @willyloman , Electorn Cash 3.3.1, Ledger Live recognises the Nano S, but not Electron Cash. Any ideas?
